Automattic-Tracks-iOS/Crash Logging/CrashLogging.swift

```diff
@@ -5,188 +5,259 @@ import CocoaLumberjack
+        guard shouldSendEvent else {
+            return nil
         }
 
-        guard let event = event else {
-            DDLogDebug("📜 Cancelling log file attachment – event is nil")
-            return shouldSendEvent
-        }
+        /// Read the current user from the Data Provider (though the Data Provider can decide not to provide it for functional or privacy reasons)
+        event.user = dataProvider.currentUser?.sentryUser
 
-        guard shouldSendEvent else {
-            DDLogDebug("📜 Cancelling log file attachment – should not send event")
-            return shouldSendEvent
+        /// Everything below this line is related to event logging, so if it's not set up we can exit
+        guard let eventLogging = self.eventLogging else {
+            DDLogDebug("📜 Cancelling log file attachment – Event Logging is not initialized")
+            return event
         }
 
         eventLogging.attachLogToEventIfNeeded(event: event)
 
-        return shouldSendEvent
-    }
-
-    /// The current state of the user's choice to opt out of data collection. Provided by the data source.
-    private static var userHasOptedOut: Bool {
-        /// If we can't say for sure, assume the user has opted out
-        guard let dataProvider = sharedInstance.dataProvider else { return true }
-        return dataProvider.userHasOptedOut
+        return event
     }
 
     /// Immediately crashes the application and generates a crash report.
-    public static func crash() {
-        Client.shared?.crash()
+    public func crash() {
+        SentrySDK.crash()
     }
 
-    public static var eventLogging: EventLogging? {
-        return sharedInstance.eventLogging
+    enum Errors: LocalizedError {
+        case unableToConstructAuthStringError
     }
 }
 
-// Manual Error Logging
+// MARK: - Manual Error Logging
 public extension CrashLogging {
 
-    /**
-     Writes the error to the Crash Logging system, and includes a stack trace.
+    ///
+    /// Writes the error to the Crash Logging system, and includes a stack trace.
+    /// - Parameters:
+    ///   - error: The error object
+    ///   - userInfo: A dictionary containing additional data about this error.
+    ///   - level: The level of severity to report in Sentry (`.error` by default)
+    func logError(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error) {
 
-     - Parameters:
-     - error: The error object
-     - userInfo: A dictionary containing additional data about this error.
-     - level: The level of severity to report in Sentry (`.error` by default)
-    */
-    static func logError(_ error: Error, userInfo: [String: Any]? = nil, level: SentrySeverity = .error) {
-        let event = Event(level: .error)
-        event.message = error.localizedDescription
+        let event = Event(level: level)
+
+        /// Use the unlocalized error message for better grouping
+        event.message = SentryMessage(formatted: (error as NSError).description)
```

Comment:
Is there a risk that the `Error` (Swift error) we're trying to load and pass to this method is NOT an `NSError` and that this could crash at runtime?

I know that the `as` (not `as!` nor `as?`) here seems to imply that the compiler seems happy with it, but I've seen cases where even if `Swift.Error` was toll-free-bridgeable from ObjC's `NSError`, if the original Swift Error was not created from an `NSError` in the first place it might not have all its fields set…

For example what happens if we have `enum MyError: Error { case foo(code: Int) }` and then call `logError(MyError.foo(code: 42), …)`? What would the `(error as NSError).description` return here?

---

EDIT: I tested it in a Playground and it seems in those cases the description is derived from the debug description, e.g. `__lldb_expr_3.MyError.foo(code: 42)`, but playgrounds are running in debug mode so wonder if that's still true in release.